lib/haml_lint/adapter.rb

```diff
@@ -0,0 +1,36 @@
+require 'haml_lint/adapter/haml_4'
+require 'haml_lint/adapter/haml_5'
+require 'haml_lint/exceptions'
+
+module HamlLint
+  # Determines the adapter to use for the current Haml version
+  class Adapter
+    # Detects the adapter to use for the current Haml version
+    #
+    # @example
+    #   HamlLint::Adapter.detect_class.new('%div')
+    #
+    # @api public
+    # @return [Class] the adapter class
+    # @raise [HamlLint::Exceptions::UnknownHamlVersion]
+    def self.detect_class
+      version = haml_version
+      case version
+      when '~> 4.0' then HamlLint::Adapter::Haml4
+      when '~> 5.0' then HamlLint::Adapter::Haml5
+      else fail HamlLint::Exceptions::UnknownHamlVersion, "Cannot handle Haml version: #{version}"
+      end
+    end
+
+    # Determines the approximate version of Haml that is installed
+    #
+    # @api private
+    # @return [String] the approximate Haml version
+    def self.haml_version
+      Gem::Version
+        .new(Haml::VERSION)
+        .approximate_recommendation
+    end
+    private_class_method :haml_version
+  end
+end
```

Comment:
This whole class is very clean. I didn't know about `approximate_recommendation`, so I'm glad you introduced it to me. It makes the case statement easy to read and understand.